src/installer/scripts.py

```diff
@@ -0,0 +1,116 @@
+import io
+import shlex
+import zipfile
+
+import six
+
+from installer import _scripts
+from installer._compat import importlib_resources
+from installer._compat.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Literal, Optional, Tuple
+
+    LauncherKind = Literal["", "t32", "t64", "t_arm", "w32", "w64", "w_arm"]
+
+
+__all__ = ["Script"]
+
+_ALLOWED_LAUNCHER_KINDS = {"", "t32", "t64", "t_arm", "w32", "w64", "w_arm"}
+
+_SCRIPT_TEMPLATE = """\
+# -*- coding: utf-8 -*-
+import re
+import sys
+from {module} import {import_name}
+if __name__ == "__main__":
+    sys.argv[0] = re.sub(r"(-script\\.pyw|\\.exe)?$", "", sys.argv[0])
+    sys.exit({func_path}())
+"""
+
+
+def _get_launcher_data(kind):
+    # type: (LauncherKind) -> Optional[six.binary_type]
+    if not kind:
+        return None
+    name = "{}.exe".format(kind)
+    return importlib_resources.read_binary(_scripts, name)
+
+
+def _is_executable_simple(executable):
+    # type: (six.binary_type) -> bool
+    if b" " in executable:
+        return False
+    shebang_length = len(executable) + 3  # Prefix #! and newline after.
+    # According to distlib, Darwin can handle up to 512 characters. But I want
+    # to avoid platform sniffing to make this as platform agnostic as possible.
+    # The "complex" script isn't that bad anyway.
+    return shebang_length <= 127
+
+
+def _quote_compat(s):  # pragma: no cover
+    # type: (six.text_type) -> six.text_type
+    """Crude implementation taken from shlex.quote().
+    """
+    return u"'" + s.replace(u"'", u"'\"'\"'") + u"'"
+
+
+def _build_shebang(executable, kind):
+    # type: (six.text_type, LauncherKind) -> six.binary_type
+    """Build a shebang line.
+
+    The non-launcher cases are taken directly from distlib's implementation,
+    which tries its best to account for command length, spaces in path, etc.
+
+    https://bitbucket.org/pypa/distlib/src/58cd5c6/distlib/scripts.py#lines-124
+    """
+    executable_bytes = six.ensure_binary(executable, encoding="utf-8")
+    if kind:  # The launcher can just use the command as-is.
+        return b"#!" + executable_bytes
+    if _is_executable_simple(executable_bytes):
+        return b"#!" + executable_bytes
+
+    quote = getattr(shlex, "quote", _quote_compat)
+    quoted = six.ensure_binary(quote(executable), encoding="utf-8")
+    # I don't understand a lick what this is trying to do.
+    return b"#!/bin/sh\n'''exec' " + quoted + b' "$0" "$@"\n' + b"' '''"
```

Comment:
Can you explain how the script is supposed to work? This seems a bit weird, the script will call the executable passing the script name as the first argument. So:

```sh
/path/to/my/executable script.sh arg1 arg2 arg3 ...
```

Also, this does not build the shebang, it builds the whole script. And I am not even sure why it is needed in the first place...

Reply:
Shebang has a path length limit.